### ISSUE TITLE
Feature importance for instance prediction

### ISSUE THREAD

**Issue by tlapusan** [parrt/dtreeviz#79]
Hi @parrt,
In another issue (#65 ), we have discussed about making a prediction more English readable. One solution was to show the feature importances only for nodes on the prediction path. 
I have looked into sklearn source code and I have found how feature importance is calculated for a decision tree. I could extract that code and adapt it to get the feature importance for a prediction path. 
I could come up with a visualisation sample in the next days.
What do you think ? :)

**Comment by parrt** [comment]
Their def of feature importance is biased though. ;) See [Beware Default Random Forest Importances](https://explained.ai/rf-importance/index.html)

**Comment by tlapusan** [comment]
> Their def of feature importance is biased though. ;) See [Beware Default Random Forest Importances](https://explained.ai/rf-importance/index.html)

Thanks for the link. I will take a look over it.

**Comment by tlapusan** [comment]
Hi @parrt, I have thought how to adapt permutation importance for instance prediction path...

1. We need to establish a "validation" set. 
This could be the samples from training set which arrived in the prediction path leaf(we already have this functionality)
Or we can extract the samples from raw validation set which arrived in the leaf. (this could be pretty time consumption)

2. Shuffle features from instance prediction path.
If we shuffle features values, we will have totally different prediction paths compare with original one. Maybe this is what we want....but I'm not pretty sure about other side effects.

3. Compare baseline validation accuracy with the new one (accuracy based on shuffling).

What is your opinion ?

**Comment by parrt** [comment]
Hmm...now that I think about this, maybe all we need is something that indicates in english what feature space boundaries are. Like you didn't get loan because your NUMCREDITCARDS was greater than 3 or something.

**Comment by tlapusan** [comment]
Sure, this is also an option to describe the prediction path.
I wouldn't abandon the idea to show also the default feature importances from instance path....they are not biased all the time :)

We can define a parameter for dtreeviz() method and let the user to chose what values to display for prediction path features. Maybe in the feature we will find a way to implement the permutation importance also.

**Comment by tlapusan** [comment]
hi @parrt, a use-case to deal with for permutation importance ...
Something to take in consideration is when training set is 'different' than validation set. For example if feature age would have a different set of values in training set compare with validation set. I think that permutation importance is biased by different values distributions in train/validation.
Of course, when we split in training and validation sets, we should pay attention to have the same distribution of values, but things happens...

**Comment by parrt** [comment]
Yes, I think it is useful to show the increase in impurity in the nodes as we go down but I don't think it's a high priority to add the feature importance derived from the tree, particularly because I'm going to publisher much better method in the next couple of months :)

The permutation importance chooses features based upon the validation set which really should be from the same distribution; if not it's time to think about why. Is it a bug or do we need to talk to the client about their data?   I have found cases where features that are very strong at prediction for training sets are not necessarily the ones good at predicting validation sets.

**Comment by tlapusan** [comment]
> now that I think about this, maybe all we need is something that indicates in english what feature space boundaries are. Like you didn't get loan because your NUMCREDITCARDS was greater than 3 or something.

Should I start thinking how to implement this one ? or do you plan to work on in in the next months ?

**Comment by tlapusan** [comment]
> > now that I think about this, maybe all we need is something that indicates in english what feature space boundaries are. Like you didn't get loan because your NUMCREDITCARDS was greater than 3 or something.
> 
> Should I start thinking how to implement this one ? or do you plan to work on in in the next months ?

HI @parrt, any feedback on this?

**Comment by parrt** [comment]
yes, please try out an "explain(x)" method or some such that tries to merge decisions made on path to leaf and explain in english.

**Comment by tlapusan** [comment]
hi @parrt, I have an initial version for prediction path explanation.

<img width="1020" alt="Screen Shot 2020-02-13 at 6 06 36 PM" src="https://user-images.githubusercontent.com/12815158/74453815-9e364980-4e8b-11ea-8909-d58319d666f3.png">

Do you think we need to add other text for method's output ?

I have created explain_prediction_path() method to be easily adaptable for other types of explanation (for the future).

**Comment by parrt** [comment]
heh, thanks looks pretty good!

**Comment by tlapusan** [comment]
> heh, thanks looks pretty good!

cool, I need to do some refactoring, after that I will make a PR.

**Comment by tlapusan** [comment]
Hi @parrt,

I have added also implementation for prediction path using sklearn default method. 
I have added into its method's documentation about the fact it can be biased (added also the link to your article). 
Taking in consideration that it is not biased all the time and that majority of sklearn users are using this version of feature importance, I think that this implementation is also useful to have implemented in dtreeviz. 
What do you think ?

<img width="884" alt="prediction_path" src="https://user-images.githubusercontent.com/12815158/75261853-581aa780-57f4-11ea-890c-239affab018a.png">

**Comment by parrt** [comment]
Looks really great!
